lib/wasi/src/fs/inode_guard.rs

```diff
@@ -237,209 +198,229 @@ impl<'a> Future for InodeValFilePollGuardJoin<'a> {
-                    Ok(a) => a,
-                    Err(e) => {
-                        error = map_io_err(e);
-                        0
-                    }
-                };
-                return Poll::Ready(Event {
-                    userdata: self.subscription.userdata,
-                    error,
-                    type_: self.subscription.type_,
-                    u: match self.subscription.type_ {
-                        Eventtype::FdRead | Eventtype::FdWrite => EventUnion {
-                            fd_readwrite: EventFdReadwrite {
-                                nbytes: bytes_available as u64,
-                                flags: if bytes_available == 0 {
-                                    Eventrwflags::FD_READWRITE_HANGUP
-                                } else {
-                                    Eventrwflags::empty()
+                            Eventtype::Clock => EventUnion { clock: 0 },
+                        },
+                    })
+                    .ok();
+                }
+                Poll::Ready(bytes_available) => {
+                    let mut error = Errno::Success;
+                    let bytes_available = match bytes_available {
+                        Ok(a) => a,
+                        Err(e) => {
+                            error = map_io_err(e);
+                            0
+                        }
+                    };
+                    ret.push(Event {
+                        userdata: self.subscription.userdata,
+                        error,
+                        type_: self.subscription.type_,
+                        u: match self.subscription.type_ {
+                            Eventtype::FdRead | Eventtype::FdWrite => EventUnion {
+                                fd_readwrite: EventFdReadwrite {
+                                    nbytes: bytes_available as u64,
+                                    flags: if bytes_available == 0 {
+                                        Eventrwflags::FD_READWRITE_HANGUP
+                                    } else {
+                                        Eventrwflags::empty()
+                                    },
                                 },
                             },
+                            Eventtype::Clock => EventUnion { clock: 0 },
                         },
-                        Eventtype::Clock => EventUnion { clock: 0 },
-                    },
-                });
-            }
+                    })
+                    .ok();
+                }
+                Poll::Pending => {}
+            };
+        }
+
+        if ret.len() > 0 {
+            return Poll::Ready(ret);
         }
         Poll::Pending
     }
 }
 
 #[derive(Debug)]
 pub(crate) struct InodeValFileReadGuard {
-    #[allow(dead_code)]
-    file: Arc<RwLock<Box<dyn VirtualFile + Send + Sync + 'static>>>,
     guard: RwLockReadGuard<'static, Box<dyn VirtualFile + Send + Sync + 'static>>,
```

Comment:
I know this isn't new, but... I  do not .. like this ... at all...

Note: `tokio::sync::RwLock` has `Owned` variants of the guards that work for `Arc<RwLock<T>>`, created with `RwLock::read/write_owned()`.
That would be much better to use...
(the tokio locks still also have sync lock methods)

Reply:
As discussed - updated here https://github.com/wasmerio/wasmer/commit/31c6eca32eeefaae422bf17c0048d1bd8b0d0300